File: benchmarks/tsb/bench_timedelta_index.ts
/**
 * Benchmark: TimedeltaIndex.fromTimedeltas / fromRange / fromStrings — TimedeltaIndex construction.
 * Outputs JSON: {"function": "timedelta_index", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Timedelta, TimedeltaIndex } from "../../src/index.ts";

const SIZE = 1_000;
const WARMUP = 5;
const ITERATIONS = 100;

const deltas = Array.from({ length: SIZE }, (_, i) =>
  Timedelta.fromComponents({ days: i, hours: i % 24 }),
);
const startTd = Timedelta.fromComponents({ days: 0 });
const stopTd = Timedelta.fromComponents({ days: SIZE });
const stepTd = Timedelta.fromComponents({ days: 1 });
const strings = Array.from({ length: SIZE }, (_, i) => `${i}D`);

for (let i = 0; i < WARMUP; i++) {
  TimedeltaIndex.fromTimedeltas(deltas);
  TimedeltaIndex.fromRange(startTd, stopTd, stepTd);
  TimedeltaIndex.fromStrings(strings);
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  TimedeltaIndex.fromTimedeltas(deltas);
  TimedeltaIndex.fromRange(startTd, stopTd, stepTd);
  TimedeltaIndex.fromStrings(strings);
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "timedelta_index",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);

File: benchmarks/pandas/bench_timedelta_index.py
"""Benchmark: pd.TimedeltaIndex construction from timedeltas/range/strings."""
import json
import time
import numpy as np
import pandas as pd

SIZE = 1_000
WARMUP = 5
ITERATIONS = 100

deltas = [pd.Timedelta(days=i, hours=i % 24) for i in range(SIZE)]
start_td = pd.Timedelta(days=0)
stop_td = pd.Timedelta(days=SIZE)
step_td = pd.Timedelta(days=1)
strings = [f"{i}D" for i in range(SIZE)]

for _ in range(WARMUP):
    pd.TimedeltaIndex(deltas)
    pd.timedelta_range(start=start_td, end=stop_td, freq=step_td)
    pd.to_timedelta(strings)

start = time.perf_counter()
for _ in range(ITERATIONS):
    pd.TimedeltaIndex(deltas)
    pd.timedelta_range(start=start_td, end=stop_td, freq=step_td)
    pd.to_timedelta(strings)
total = (time.perf_counter() - start) * 1000

print(json.dumps({"function": "timedelta_index", "mean_ms": total / ITERATIONS, "iterations": ITERATIONS, "total_ms": total}))
